### PR TITLE
swarm/network/stream: refactor bootstrapping of tests

### PR DESCRIPTION
**Pull request by nonsense**
After reviewing @holisticode 's PR https://github.com/ethereum/go-ethereum/pull/18972 and discussing my opinion on the test it was adding, I looked into refactoring the service bootstrap that is replicated over 10 simulation tests.

I came up with 3 constructors that pretty much abstract the functionality we need:

```
// newNetStoreAndDelivery is a default constructor for BzzAddr, NetStore and Delivery, used in Simulations

// newNetStoreAndDeliveryWithBzzAddr is a constructor for NetStore and Delivery, used in Simulations, accepting any BzzAddr

// newNetStoreAndDeliveryWithRequestFunc is a constructor for NetStore and Delivery, used in Simulations, accepting any NetStore.RequestFunc
```

These reduce the copy-pasted boilerplate that we have in the simulation tests.

I didn't get as far as reviewing the constructor for `NewRegistry`, but maybe there are things we could simplify there as well.

Let me know what you think.

### CODE REVIEW

**Comment by janos** [comment]
> Note that there were number of places where we were not calling `r.Close()` in the clean up. Not sure if this is an issue or is fine, @janos probably knows best.

Good catch @nonsense. I thought that network shutdown will do that, but actually it will just call Stop method on Service (in this case Registry). Registry closing is required for closing intervals database.

File: swarm/network/stream/common_test.go [hunk 1]
```diff
@@ -66,6 +69,80 @@ func init() {
...
+	netStore.NewNetFetcherFunc = network.NewFetcherFactory(rf, true).New
+
+	return addr, netStore, delivery, cleanup, nil
+}
+
+func netStoreAndDeliveryWithAddr(ctx *adapters.ServiceContext, bucket *sync.Map, addr *network.BzzAddr) (*storage.NetStore, *Delivery, func(), error) {
+	n := ctx.Config.Node()
+
+	store, datadir, err := createTestLocalStorageForID(n.ID(), addr)
+	if *useMockStore {
```
Comment (nonsense) [inline swarm/network/stream/common_test.go:116]: @janos I am pretty sure this doesn't work, but haven't tried it just yet. Not sure how we cast a mock store to `*storage.LocalStore` further down, but this is how it is in the current code as well.

Comment (janos) [reply swarm/network/stream/common_test.go:116]: We have found that mock store in stream tests were broken after some rewrite. I am not sure if they were fixed. Maybe @holisticode have more details, but I will have a look also.

Comment (janos) [reply swarm/network/stream/common_test.go:116]: But it is important that mock store is not casted to LocalStore, there were numerous discussion on this, but injected down to ldbstore.

Comment (nonsense) [reply swarm/network/stream/common_test.go:116]: At least now we have to fix it at one place, and not at multiple :)

**Review by janos — APPROVED** [review]
Great generalization. LGTM.

**Review by holisticode — COMMENTED** [review]
Great PR, simplifies simulation creation a lot.

@Janos @nonsense would it actually make sense to move these helper functions into the `simulation` package itself instead of here? In other words, could they be helpful outside the `stream` package?

**Review by holisticode — APPROVED** [review]

**Review by zelig — APPROVED** [review]

**Comment by zelig** [comment]
merged as https://github.com/ethereum/go-ethereum/pull/18975

**Pull request closed without merging**